bitcoin/src/bip32.rs

```diff
@@ -439,17 +498,24 @@ impl FromStr for NormalDerivationPath {
         let path = path.strip_prefix("m/").unwrap_or(path);
 
         let parts = path.split('/');
-        let ret: Result<Vec<ChildNumber>, Error> = parts.map(str::parse).collect();
-        Ok(ret?.as_slice().try_into()?)
+        let ret: Result<Vec<NormalChildNumber>, Error> = parts.map(str::parse).collect();
+        Ok(NormalDerivationPath::from(ret?))
     }
 }
 
 impl IntoIterator for NormalDerivationPath {
     type Item = NormalChildNumber;
-    type IntoIter = std::vec::IntoIter<NormalChildNumber>;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
     fn into_iter(self) -> Self::IntoIter { self.0.into_iter() }
 }
 
+impl<'a> IntoIterator for &'a NormalDerivationPath {
+    type Item = &'a NormalChildNumber;
+    type IntoIter = std::slice::Iter<'a, NormalChildNumber>;
```

Comment:
LIke ?

```Rust
impl<'a> IntoIterator for &'a NormalDerivationPath {
    type Item = dyn std::iter::IteratorItem<NormalChildNumber = &'a NormalChildNumber>;
    type IntoIter = Box<dyn Iterator<Item = &'a NormalChildNumber>>;

    fn into_iter(self) -> Self::IntoIter {
        Box::new(self.0.iter())
    }
}
```

Reply:
No, just create a newtypes for the iterators (Using standard names `IntoIter` and `Iter`, need to be moved in a separate module though.)

Reply:
```Rust
mod iter {
    use super::*;
    // An iterator over a [`NormalDerivationPath`]
    pub struct NormalDerivationPathIter<'a>(std::slice::Iter<'a, NormalChildNumber>);
    impl<'a> Iterator for NormalDerivationPathIter<'a> {
        type Item = &'a NormalChildNumber;

        fn next(&mut self) -> Option<Self::Item> { self.0.next() }
    }
    impl<'a> IntoIterator for &'a NormalDerivationPath {
        type Item = &'a NormalChildNumber;
        type IntoIter = NormalDerivationPathIter<'a>;

        fn into_iter(self) -> Self::IntoIter { NormalDerivationPathIter(self.0.iter()) }
    }
}
```

Like this ? Wouldnt it be confusing ? because we have a iterator struct for `DerivationPath` but for its children.
If this kind on newtype that you asking for, would be good to make the same for `DerivationPath`(?).(A Iterator over the childrhen of `NormalDerivationPath` is not needed because we can easily convert to `DerivationPathIterator`)

Reply:
I mean

```rust
pub mod derivation_path {
    pub struct DerivationPath(Vec<ChildNumber>);

    pub struct Iter<'a>(core::slice::Iter<'a, ChildNumber>);

    impl<'a> Iterator for Iter<'a> {
        type Item = ChildNumber;

        fn next(&mut self) -> Option<Self::Item> {
            self.0.next().copied()
        }
    }

    // just directly delegate the Iterator trait
    pub struct IntoIter(alloc::vec::IntoIter<ChildNumber>);

    // IntoIterator impl pretty much like you wrote.
}
```

And same for `NormalDerivationPath`.

It won't be confusing because it's idiomatic Rust. The entire `std` and other quality crates use API like this.

`derive_path` definitely needs to accept generic iterator.